File: frontend/src/types/feedback.ts
export const Category = {
  BUG: 'bug',
  FEATURE: 'feature',
  IMPROVEMENT: 'improvement',
  COMPLAINT: 'complaint',
  PRAISE: 'praise',
  OTHER: 'other',
} as const;
export type Category = typeof Category[keyof typeof Category];

export const Sentiment = {
  POSITIVE: 'positive',
  NEGATIVE: 'negative',
  NEUTRAL: 'neutral',
  MIXED: 'mixed',
} as const;
export type Sentiment = typeof Sentiment[keyof typeof Sentiment];

export const Priority = {
  LOW: 'low',
  MEDIUM: 'medium',
  HIGH: 'high',
  CRITICAL: 'critical',
} as const;
export type Priority = typeof Priority[keyof typeof Priority];

export const Source = {
  WEB: 'web',
  EMAIL: 'email',
  API: 'api',
  SOCIAL: 'social',
  SUPPORT: 'support',
} as const;
export type Source = typeof Source[keyof typeof Source];

export const Status = {
  NEW: 'new',
  IN_REVIEW: 'in_review',
  ADDRESSED: 'addressed',
  DISMISSED: 'dismissed',
} as const;
export type Status = typeof Status[keyof typeof Status];

export interface Feedback {
  id: string;
  text: string;
  rating: number;
  category: Category;
  sentiment: Sentiment;
  priority: Priority;
  source: Source;
  status: Status;
  created_at: string;
  updated_at: string;
}

export interface FeedbackCreate {
  text: string;
  rating: number;
  category: Category;
  sentiment: Sentiment;
  priority: Priority;
  source: Source;
}

export interface FeedbackUpdate {
  text?: string;
  rating?: number;
  category?: Category;
  sentiment?: Sentiment;
  priority?: Priority;
  source?: Source;
  status?: Status;
}

export interface FeedbackListResponse {
  items: Feedback[];
  total: number;
  page: number;
  size: number;
}

export interface FeedbackStats {
  total_count: number;
  average_rating: number;
  sentiment_distribution: Record<string, number>;
  category_distribution: Record<string, number>;
  priority_distribution: Record<string, number>;
  status_distribution: Record<string, number>;
  rating_distribution: Record<string, number>;
  trends: any[];
  priority_heatmap: Record<string, Record<string, number>>;
  recent_feedback: Feedback[];
}

// Intelligence Engine types
export interface SentimentResult {
  label: string;
  score: number;
  confidence: number;
  positive_words: string[];
  negative_words: string[];
}

export interface CategoryResult {
  category: string;
  confidence: number;
  matched_keywords: string[];
}

export interface PriorityResult {
  priority: string;
  score: number;
  reasons: string[];
}

export interface AnalysisResult {
  cleaned_text: string;
  sentiment: SentimentResult;
  category: CategoryResult;
  priority: PriorityResult;
  processing_steps: string[];
}

export interface SmartFeedbackCreate {
  text: string;
  rating: number;
  source: string;
}

File: backend/app/schemas/feedback.py
"""
Feedback Pydantic schemas.

Defines request/response models and enums for the feedback domain,
including create, update, response, list, and statistics schemas.
"""

from datetime import datetime
from enum import Enum
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


# ---------------------------------------------------------------------------
# Enums
# ---------------------------------------------------------------------------

class CategoryEnum(str, Enum):
    """Valid feedback categories."""

    BUG = "bug"
    FEATURE = "feature"
    IMPROVEMENT = "improvement"
    COMPLAINT = "complaint"
    PRAISE = "praise"
    OTHER = "other"


class SentimentEnum(str, Enum):
    """Valid sentiment classifications."""

    POSITIVE = "positive"
    NEGATIVE = "negative"
    NEUTRAL = "neutral"
    MIXED = "mixed"


class PriorityEnum(str, Enum):
    """Valid priority levels."""

    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"


class SourceEnum(str, Enum):
    """Valid feedback sources."""

    WEB = "web"
    EMAIL = "email"
    API = "api"
    SOCIAL = "social"
    SUPPORT = "support"


class StatusEnum(str, Enum):
    """Valid feedback statuses."""

    NEW = "new"
    IN_REVIEW = "in_review"
    ADDRESSED = "addressed"
    DISMISSED = "dismissed"


# ---------------------------------------------------------------------------
# Request schemas
# ---------------------------------------------------------------------------

class FeedbackCreate(BaseModel):
    """Schema for creating a new feedback entry."""

    text: str = Field(..., min_length=1, description="The feedback text content.")
    rating: int = Field(..., ge=1, le=5, description="Rating from 1 to 5.")
    category: CategoryEnum = Field(..., description="Feedback category.")
    sentiment: SentimentEnum = Field(..., description="Sentiment classification.")
    priority: PriorityEnum = Field(..., description="Priority level.")
    source: SourceEnum = Field(..., description="Source of the feedback.")

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "text": "The new dashboard is really intuitive and fast!",
                "rating": 5,
                "category": "praise",
                "sentiment": "positive",
                "priority": "low",
                "source": "web",
            }
        }
    )


class FeedbackUpdate(BaseModel):
    """Schema for updating an existing feedback entry. All fields are optional."""

    text: str | None = Field(None, min_length=1, description="The feedback text content.")
    rating: int | None = Field(None, ge=1, le=5, description="Rating from 1 to 5.")
    category: CategoryEnum | None = Field(None, description="Feedback category.")
    sentiment: SentimentEnum | None = Field(None, description="Sentiment classification.")
    priority: PriorityEnum | None = Field(None, description="Priority level.")
    source: SourceEnum | None = Field(None, description="Source of the feedback.")
    status: StatusEnum | None = Field(None, description="Feedback status.")

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "status": "in_review",
                "priority": "high",
            }
        }
    )


# ---------------------------------------------------------------------------
# Response schemas
# ---------------------------------------------------------------------------

class FeedbackResponse(BaseModel):
    """Schema for a single feedback response."""

    id: UUID
    text: str
    rating: int
    category: CategoryEnum
    sentiment: SentimentEnum
    priority: PriorityEnum
    source: SourceEnum
    status: StatusEnum
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class FeedbackListResponse(BaseModel):
    """Paginated list response for feedback entries."""

    items: list[FeedbackResponse]
    total: int
    page: int
    size: int


class FeedbackStats(BaseModel):
    """Aggregated statistics for all feedback entries."""

    total_count: int
    average_rating: float
    sentiment_distribution: dict[str, int]
    category_distribution: dict[str, int]
    priority_distribution: dict[str, int]
    status_distribution: dict[str, int]
    rating_distribution: dict[str, int]
    trends: list[dict]
    priority_heatmap: dict[str, dict[str, int]]
    recent_feedback: list[FeedbackResponse]
